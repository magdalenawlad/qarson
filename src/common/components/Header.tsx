import React from 'react';
import { ShoppingCartSharp } from '@mui/icons-material';
import { useSelector } from 'react-redux';
import { RootState } from '../../app/store';
import logo from '../../assets/images/logo.png';

export const Header = () => {
  const selectedCarsCounter = useSelector((state: RootState) => state.offers.selected).length;
  return (
    <header>
      <div className='logo'>
        <img src={logo} alt='Qarson' />
      </div>
      <div className='icon'>
        <ShoppingCartSharp fontSize='large' />
        {!!selectedCarsCounter && <span className="counter">{selectedCarsCounter}</span>}
        </div>
    </header>
  );
};
