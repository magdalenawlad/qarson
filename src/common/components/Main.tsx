import React from 'react';

export const Main = ({ children }: { children: React.ReactNode }) => (
  <main>
    <div className='main-container'>
      {children}
    </div>
  </main>
);
