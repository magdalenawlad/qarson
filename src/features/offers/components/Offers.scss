@use '../../../styles/variables' as v;

.offers-container {
  width: 100%;

  .offers-label {

    span {
      font-family: 'LatoBlack';
    }
  }

  .offers-list {
    padding: 2rem 0;;
    display: grid;
    grid-template-columns: repeat(auto-fill, 300px);
    grid-gap: 2rem;
    justify-content: center;
  }
}

.offer-card {
  background: transparent;
  border-radius: 1rem;;
  box-shadow: 3px 3px 10px 2px rgba(v.$textPrimary, 0.3);
  display: flex;
  align-items: center;
  flex-direction: column;
  overflow: hidden;
  position: relative;
  @include v.shadow();

  img {
    width: auto;
    width: 100%;
    height: 168px;
  }

  &__description {
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    width: 100%;
    padding: 1rem;

    &__name {
      font-family: 'LatoBlack';
    }
  }

  &__footer {
    display: flex;
    width: 100%;
  
    > div, > button {
      flex: 1;
    }

    .offer-card__footer__button {
      display: flex;
      justify-content: center;
      align-items: center;
      background: v.$textPrimary;
      color: v.$background;
      height: 3rem;
      border: none;
      border-radius: 1rem 0 0;
      font-size: v.$fontSizeRegular;
      cursor: pointer;

      &--selected {
        background: v.$qarsonPrimary;
        color: v.$textPrimary;
        border-radius: 1rem 0 0;
        box-shadow: inset 7px 7px 14px #6e872a, inset -7px -7px 14px #d6ff52;
      }

      &:disabled {
        opacity: 0.4;
        cursor: default;
      }

      svg {
        margin-right: 0.3rem;
      }
    }

    &__price {
      display: flex;
      align-items: center;
      justify-content: center;
      font-family: 'LatoBlack';
    }
  }
  
  .indicator {
    margin: 0;
    opacity: 0.4;
    padding: 0;
    background: v.$textPrimary;
    color: v.$background;
    padding: 0.5rem 0;
    position: absolute;
    top:0;
    right:0;
    transform: translateX(30%) translateY(0%) rotate(45deg);
    transform-origin: top left;
  }

  .indicator:before,
  .indicator:after {
    content: '';
    position: absolute;
    top:0;
    margin: 0 -1px;
    width: 100%;
    height: 100%;
    background: v.$textPrimary;
  }

  .indicator:before {
    right:100%;
  }
  
  .indicator:after {
    left:100%;
  }

  .car-placeholder {
    padding: 3rem;
    opacity: 0.5
  }
}