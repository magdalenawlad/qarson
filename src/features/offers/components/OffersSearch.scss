@use '../../../styles/variables' as v;

.search {
  padding: 2rem 0;
  border-radius: 10px;
  width: 100%;

  .search-textfield {
    @include v.shadow(0.5rem);

    label.Mui-focused {
      color: v.$textPrimary;
    }

    & .MuiOutlinedInput-root {
      & fieldset {
        border: none;
      }
    }
  }

  .search-quick-filters {
    &__make {
      padding: 1rem 0 0.6rem;
      display: flex;
      flex-wrap: wrap;

      &__item {
        margin: 0 0.3rem 0.3rem 0;
        display: flex;
        align-items: center;
        justify-content: center;
        border-radius: 0.5rem;
        height: 2.6rem;
        width: 2.6rem;
        padding: 0.3rem;
        cursor: pointer;

        &--selected {
          background: v.$qarsonPrimary;
          @include v.shadow(0.5rem, v.$qarsonGreen);
        }

        img {
          height: auto;
          max-height: 100%;
          width: auto;
          max-width: 100%;
        }
      }
    }
  }

  .availability-filter {
    display: flex;
    align-items: center;

    &__checkbox {
      &.MuiCheckbox-root {
        padding: 0;
        
        &.Mui-checked {
          color: v.$qarsonPrimary;
          @include v.shadow();
  
          svg {
            fill: v.$qarsonPrimary;
            background: v.$textPrimary;
            border-radius: 0.5rem;
          }
        }
  
        svg {
          height: 2.6rem;
          width: 2.6rem;
          fill: v.$textPrimary;
        }
      }
    }

    &__label {
      padding: 0 0.3rem;
    }
  }
}