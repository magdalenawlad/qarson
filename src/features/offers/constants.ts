import vwLogo from '../../assets/images/carLogos/vw.svg';
import seatLogo from '../../assets/images/carLogos/seat.svg';
import renaultLogo from '../../assets/images/carLogos/renault.png';
import peugeotLogo from '../../assets/images/carLogos/peugeot.png';
import kiaLogo from '../../assets/images/carLogos/kia.png';
import skodaLogo from '../../assets/images/carLogos/skoda.png';

type EnumDictionary<T extends string | symbol | number, U> = {
  [K in T]: U;
};

export enum Make {
  Volkswagen = 'Volkswagen',
  Skoda = 'Skoda',
  Seat = 'Seat',
  Renualt ='Renualt',
  Pegeuot = 'Pegeuot',
  Kia = 'Kia'
}

export const MAKE_LOGO_PATH: EnumDictionary<Make, string> = {
  [Make.Volkswagen]: vwLogo,
  [Make.Skoda]: skodaLogo,
  [Make.Seat]: seatLogo,
  [Make.Renualt]: renaultLogo,
  [Make.Pegeuot]: peugeotLogo,
  [Make.Kia]: kiaLogo
};
