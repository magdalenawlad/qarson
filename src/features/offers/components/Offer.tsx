import React, { useCallback } from 'react';
import AddShoppingCartIcon from '@mui/icons-material/AddShoppingCart';
import RemoveShoppingCartIcon from '@mui/icons-material/RemoveShoppingCart';
import { useDispatch } from 'react-redux';
import autoIcon from '../../../assets/images/car-placeholder.png';
import { Make } from '../constants';
import { addToCart, removeFromCart, Offer as OfferType } from '../offersSlice';
import './Offers.scss';

type Props = OfferType & { selected: boolean };

export const Offer = ({
  id,
  make,
  model,
  engine,
  availability,
  photo,
  selected
}: Props) => {
  const dispatch = useDispatch();

  const handleCartButtonClick = useCallback((id: number) => 
    selected ? dispatch(removeFromCart(id)) : dispatch(addToCart(id)), [selected]);
 
  return (
    <div
      className={`offer-card ${!availability ? 'offer-card--disabled' : ''}`}
      data-testid='offer-card'
    >
      {!availability && <div className="indicator">Unavailable</div>}
      {photo ? (
        <img data-testid="offer-photo" src={photo} alt="Auto" />
      ) : (
        <img data-testid="offer-photo" className="car-placeholder" src={autoIcon} alt="Auto" />
      )}
      <div className='offer-card__description'>
        <div className='offer-card__description__name'>{`${Make[make]} ${model}`}</div>
        <div>{engine}</div>
      </div>
      <div className='offer-card__footer'>
        <div className='offer-card__footer__price'>$559.99</div>
        <button
          disabled={!availability}
          className={`offer-card__footer__button ${selected ? 'offer-card__footer__button--selected': ''}`}
          onClick={() => handleCartButtonClick(id)}
        >
          {selected ? <RemoveShoppingCartIcon /> : <AddShoppingCartIcon />}
          {selected ? 'Remove' : 'Add'}
        </button>
      </div>
    </div>
  );
}
