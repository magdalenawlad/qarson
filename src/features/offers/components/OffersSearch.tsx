import React, { useCallback } from 'react';
import { TextField } from '@mui/material';
import Checkbox from '@mui/material/Checkbox';
import SearchIcon from '@mui/icons-material/Search';
import { InputAdornment } from '@mui/material';
import { useDispatch, useSelector } from 'react-redux';
import { AppDispatch, RootState } from '../../../app/store';
import {
  changeTextFilter,
  changeMakeQuickFilter,
  changeShowOnlyAvailableFilter
} from '../offersSlice';
import { getMakeLogoPath } from '../helpers';
import { Make } from '../constants';
import './OffersSearch.scss';

export const OffersSearch = () => {
  const dispatch = useDispatch<AppDispatch>();
  const isLoading = useSelector((state: RootState) => state.offers.isLoading);
  const textFilter = useSelector((state: RootState) => state.offers.filters.text);
  const onlyAvailableFilter = useSelector((state: RootState) => state.offers.filters.showOnlyAvailable);
  const quickMakeFilters = useSelector((state: RootState) => state.offers.filters.make);

  const handleTextFilterChange = useCallback((event: React.ChangeEvent<HTMLInputElement>) =>
    dispatch(changeTextFilter(event.target.value)), []);

  const handleMakeQuickFilterChange = useCallback((name: Make, selected: boolean) =>
    dispatch(changeMakeQuickFilter({ name, selected })), []);

  const handleOnlyAvailableFilterChange = useCallback(() => {
    dispatch(changeShowOnlyAvailableFilter())}, []);

  return (
    <div className='search'>
      <div>
        <TextField
          fullWidth
          id='quick-filter-text'
          variant='outlined'
          value={textFilter}
          onChange={handleTextFilterChange}
          className='search-textfield'
          InputProps={{
            endAdornment:
              <InputAdornment position="end">
                <SearchIcon />
              </InputAdornment>
          }}
        />
      </div>
      <div className='search-quick-filters'>
        {quickMakeFilters && !!quickMakeFilters.length && (
          <div
            className='search-quick-filters__make'
            data-testid='search-quick-filters-make'
          >
            {quickMakeFilters.map(({ name, selected }) => (
              <div
                key={name}
                className={
                  `search-quick-filters__make__item ${selected ? 'search-quick-filters__make__item--selected': ''}`
                }
                data-testid='search-quick-filters-make__item'
                onClick={() => handleMakeQuickFilterChange(name, !selected)}
              >
                <img src={getMakeLogoPath(name)} />
              </div>
              )
            )}
          </div>
        )}
        {!isLoading && (
          <div className='availability-filter'>
            <Checkbox
              size='medium'
              className='availability-filter__checkbox'
              value={onlyAvailableFilter}
              onChange={handleOnlyAvailableFilterChange}
            />
            <div className='availability-filter__label'>Available now</div>
          </div>
        )}
      </div>
    </div>
  );
}
