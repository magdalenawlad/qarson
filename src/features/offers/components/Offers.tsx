import React from 'react';
import { useSelector } from 'react-redux';
import { RootState } from '../../../app/store';
import { Offer } from './Offer';
import { OffersSearch } from './OffersSearch';
import './Offers.scss';
import { Make } from '../constants';

export const Offers = () => {
  const data = useSelector((state: RootState) => state.offers.data);
  const selectedOffers = useSelector((state: RootState) => state.offers.selected);
  const filters = useSelector((state: RootState) => state.offers.filters);
  const isLoading = useSelector((state: RootState) => state.offers.isLoading);

  const filteredData = () => {
    let result = data;
    const makeFiltersApplied = filters.make.find(({ selected }) => selected);

    if (makeFiltersApplied) {
      result = result.filter(({ make }) => {
        const searched = filters
          .make.find((({ name }) => {
            return name === make;
          }));

        return searched?.selected
      });
    }

    if (filters.text) {
      const { text } = filters;
      result = result.filter(({ make, model }) =>
        Make[make].toLowerCase().includes(text.toLowerCase())
          || model.toLowerCase().includes(text.toLowerCase()));
    }

    if (filters.showOnlyAvailable) {
      result = result.filter(({ availability }) => !!availability);
    }

    return result;
  };

  const carsToShow = filteredData();

  return (
    <>
      <OffersSearch />
      <div className='offers-container'>
        {isLoading ? (
          <div>Loading...</div>
        ): (
          <div className='offers'>
            {carsToShow && (
              <div className='offers-label'>
                Offers found: <span>{carsToShow.length}</span>
              </div>
            )}
            <div className='offers-list'>
              {carsToShow && carsToShow.map((item) => (
                <Offer
                  key={`item-${item.id}`}
                  selected={selectedOffers.indexOf(item.id) >=0}
                  {...item}
                />
              ))}
            </div>
          </div>
        )}
      </div>
    </>
  )
};
