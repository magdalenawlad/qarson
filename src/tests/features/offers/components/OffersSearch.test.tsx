import { screen, within } from '@testing-library/react';
import { renderWithProviders, initialState } from '../../../test-utils';
import { OffersSearch } from '../../../../features/offers/components/OffersSearch';
import { Make } from '../../../../features/offers/constants';

test('renders correctly', () => {
  const { container } = renderWithProviders(<OffersSearch />);

  expect(container).toMatchSnapshot();
});

test('renders correctly make quick filters', () => {
  const preloadedState = {
    ...initialState,
    offers: {
      ...initialState.offers,
      filters: {
        ...initialState.offers.filters,
        make: [
          {
            name: Make.Volkswagen,
            selected: false
          },
          {
            name: Make.Skoda,
            selected: false
          },
          {
            name: Make.Seat,
            selected: false
          }
        ]
      }
    }
  }

  renderWithProviders(<OffersSearch />, { preloadedState });

  const makeQuickFilters = screen.getByTestId('search-quick-filters-make');

  expect(makeQuickFilters).toBeInTheDocument();
  expect(within(makeQuickFilters)
    .queryAllByTestId('search-quick-filters-make__item'))
    .toHaveLength(3);
});

test('renders make quick filter with correct class when selected', () => {
  const preloadedState = {
    ...initialState,
    offers: {
      ...initialState.offers,
      filters: {
        ...initialState.offers.filters,
        make: [
          {
            name: Make.Volkswagen,
            selected: true
          }
        ]
      }
    }
  }

  renderWithProviders(<OffersSearch />, { preloadedState });

  const makeQuickFilters = screen.getByTestId('search-quick-filters-make');

  expect(within(makeQuickFilters)
    .queryByTestId('search-quick-filters-make__item'))
    .toHaveClass('search-quick-filters__make__item--selected');
});

test('renders availability filter unchecked when filter not applied', () => {
  const preloadedState = {
    ...initialState,
    offers: {
      ...initialState.offers,
      filters: {
        ...initialState.offers.filters,
        showOnlyAvailable: false
      }
    }
  }

  const { container } = renderWithProviders(<OffersSearch />, { preloadedState });
  const availabilityCheckbox =
    container.querySelector('.availability-filter__checkbox > input');

  expect(availabilityCheckbox).toHaveProperty('value', 'false');
});

test('renders availability filter unchecked when filter applied', () => {
  const preloadedState = {
    ...initialState,
    offers: {
      ...initialState.offers,
      filters: {
        ...initialState.offers.filters,
        showOnlyAvailable: true
      }
    }
  }

  const { container } = renderWithProviders(<OffersSearch />, { preloadedState });
  const availabilityCheckbox =
    container.querySelector('.availability-filter__checkbox > input');

  expect(availabilityCheckbox).toHaveProperty('value', 'true');
});

test('renders text filter with correct value when text applied', () => {
  const preloadedState = {
    ...initialState,
    offers: {
      ...initialState.offers,
      filters: {
        ...initialState.offers.filters,
        text: 'Sample search string'
      }
    }
  }
  
  const { container } = renderWithProviders(<OffersSearch />, { preloadedState });
  const input = container.querySelector('#quick-filter-text');

  expect(input).toHaveValue('Sample search string');
});