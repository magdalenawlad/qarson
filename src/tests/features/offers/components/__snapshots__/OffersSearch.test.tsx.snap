// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    class="search"
  >
    <div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root search-textfield css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            id="quick-filter-text"
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SearchIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-hdw1oc"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="search-quick-filters"
    >
      <div
        class="availability-filter"
      >
        <span
          class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary availability-filter__checkbox css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
            value="false"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxOutlineBlankIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <div
          class="availability-filter__label"
        >
          Available now
        </div>
      </div>
    </div>
  </div>
</div>
`;
